src/Fhir.Anonymizer.Shared.Core/Utility/SubstituteUtility.cs

```diff
@@ -0,0 +1,126 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using Fhir.Anonymizer.Core.Models;
+using Fhir.Anonymizer.Core.Processors;
+using Hl7.Fhir.ElementModel;
+using Hl7.Fhir.Specification;
+
+namespace Fhir.Anonymizer.Core.Utility
+{
+    public class SubstituteUtility
+    {
+        private static readonly PocoStructureDefinitionSummaryProvider s_provider = new PocoStructureDefinitionSummaryProvider();
+
+        public static ProcessResult SubstituteNode(ElementNode node, ElementNode replacementNode, HashSet<ElementNode> visitedNodes, HashSet<ElementNode> keepNodes)
+        {
+            var processResult = new ProcessResult();
+            if (node == null || replacementNode == null || visitedNodes.Contains(node))
+            {
+                return processResult;
+            }
+
+            // children names to replace, multiple to multiple replacement
+            var replaceChildrenNames = replacementNode.Children().Select(element => element.Name).ToHashSet();
+            foreach (var name in replaceChildrenNames)
+            {
+                var children = node.Children(name).Cast<ElementNode>().ToList();
+                var targetChildren = replacementNode.Children(name).Cast<ElementNode>().ToList();
+
+                int i = 0;
+                foreach (var child in children)
+                {
+                    if (visitedNodes.Contains(child))
+                    {
+                        // Skip replacement if child already processed before.
+                        i++;
+                        continue;
+                    }
+                    else if (i < targetChildren.Count)
+                    {
+                        // We still have target nodes, do replacement
+                        processResult.Update(SubstituteNode(child, targetChildren[i ++], visitedNodes, keepNodes));
+                    }
+                    else if (keepNodes.Contains(child))
+                    {
+                        // Substitute with an empty node when no target node available but we need to keep this node
+                        processResult.Update(SubstituteNode(child, GetDummyNode(), visitedNodes, keepNodes));
+                    }
+                    else
+                    {
+                        // Remove source node when no target node availabe and we don't need to keep the source node
+                        node.Remove(child);
+                        processResult.AddProcessRecord(AnonymizationOperations.Substitute, child);
```

Comment:
For result, not sure why we need to record all child?

Reply:
Thanks. I have removed the records for children nodes.